### PR TITLE
api: Add getHeader api and blockNumberOrHashType

### PR DESCRIPTION
**Pull request by hyochan-brown**
## Proposed changes
To enhance compatibility with Ethereum([#19491](https://github.com/ethereum/go-ethereum/pull/19491)).
- Add BlockNumberOrHash Type & test code 
    - BlockNumber
    - BlockHash
    - RequireCanonical(just for compatibility. not using now)
```
curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"debug_getBlockRlp","params":[88000],"id":1}' http://localhost:8551
curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"debug_getBlockRlp","params":["0xf20b03cbcde40f7b0a0055bb1aeaf3c13efc73a67f20eaf58fd4fafc6139257b"],"id":1}' http://localhost:8551
{"jsonrpc":"2.0","id":1,"result": "f90713f9070fa0dd36161bb4d8616e53bfc82e4554332668c297d69cd693 ... " } // same result
```
    
- Change param type BlockNumber or BlockHash to BlockNumberOrHash  if the param is not specified in the method name.
- Add getHeader api
    - getHeaderByNumber
    - getHeaderByHash
    - klay.getHeader
    
```
 curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"klay_getHeaderByNumber","params":[70000],"id":1}' http://localhost:8551
{"jsonrpc":"2.0","id":1,"result":{"parentHash":"0x3219cf9229e5657c3e57869981ed7a382ac75714f134ddafb18d373e01336fb0","reward":"0x5f1dbd747996d8d31e2ab0317be7ffffd155522a", ... }
```
```
> klay.getHeader(160000)
{
  blockScore: "0x1",
  extraData: ...
  ...
}
```

- Add integer-decimal check in blockNumber unmarshalJson
```
 curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"klay_getHeaderByNumber","params":[70000],"id":1}' http://localhost:8551
 curl -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"klay_getHeaderByNumber","params":["0x11170"],"id":1}' http://localhost:8551
{"jsonrpc":"2.0","id":1,"result": "blockScore": "0x1", ... } // same result
```

 
 
## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

### CODE REVIEW

**Review by ehnuje — DISMISSED** [review]

File: api/api_public_blockchain.go [hunk 1]
```diff
@@ -150,6 +150,34 @@ func (s *PublicBlockChainAPI) GetAccount(ctx context.Context, address common.Add
 	return serAcc, state.Error()
 }
 
+// GetHeaderByNumber returns the requested canonical block header.
+// * When blockNr is -1 the chain head is returned.
+// * When blockNr is -2 the pending chain head is returned.
+func (s *PublicBlockChainAPI) GetHeaderByNumber(ctx context.Context, number rpc.BlockNumber) (map[string]interface{}, error) {
```
Comment (jeongkyun-oh) [inline api/api_public_blockchain.go:156]: @hyochan-brown I think we need to update `console/web3ext/web3ext.go` file as well.

File: api/api_public_blockchain.go [hunk 2]
```diff
@@ -150,6 +150,34 @@ func (s *PublicBlockChainAPI) GetAccount(ctx context.Context, address common.Add
...
+	header, err := s.b.HeaderByNumber(ctx, number)
+	if header != nil && err == nil {
+		response, err := s.rpcOutputHeader(header)
+		if number == rpc.PendingBlockNumber {
```
Comment (jeongkyun-oh) [inline api/api_public_blockchain.go:160]: Unlike ethereum, we do not support pending block number. So, I think this if clause is not needed. (HeaderByNumber will return an error `ErrPendingBlockNotSupported`.

File: console/web3ext/web3ext.go [hunk 3]
```diff
@@ -716,6 +716,17 @@ web3._extend({
 			params: 2,
 			inputFormatter: [web3._extend.formatters.inputAddressFormatter, web3._extend.formatters.inputDefaultBlockNumberFormatter],
 		}),
+		new web3._extend.Method({
```
Comment (jeongkyun-oh) [inline console/web3ext/web3ext.go:719]: `klay_getBlock` supports a number and a hash as well. How about adding `klay_getHeader`?

```
var blockCall = function (args) {
    return (utils.isString(args[0]) && args[0].indexOf('0x') === 0) ? "klay_getBlockByHash" : "klay_getBlockByNumber";
};

var getBlock = new Method({
    name: 'getBlock',
    call: blockCall,
    params: 2,
    inputFormatter: [formatters.inputBlockNumberFormatter, function (val) { return !!val; }],
    outputFormatter: formatters.outputBlockFormatter
});
```

File: networks/rpc/types_test.go [hunk 4]
```diff
@@ -24,6 +24,8 @@ import (
 	"encoding/json"
 	"testing"
 
+	"github.com/klaytn/klaytn/common"
+
 	"github.com/klaytn/klaytn/common/math"
```
Comment (yoomee1313) [inline networks/rpc/types_test.go:29]: There's an enter between two lines.
```suggestion
	"github.com/klaytn/klaytn/common"
	"github.com/klaytn/klaytn/common/math"
```

File: networks/rpc/types.go [hunk 5]
```diff
@@ -170,3 +172,94 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
...
+}
+
+func (bnh *BlockNumberOrHash) Hash() (common.Hash, bool) {
+	if bnh.BlockHash != nil {
+		return *bnh.BlockHash, true
+	}
+	return common.Hash{}, false
+}
+
+func BlockNumberOrHashWithNumber(blockNr BlockNumber) BlockNumberOrHash {
```
Comment (yoomee1313) [inline networks/rpc/types.go:251]: What about adding New at `BlockNumberOrHashWithNumber` and `BlockNumberOrHashWithHash`? If you think this is not necessary, you don't have to apply
```suggestion
func NewBlockNumberOrHashWithNumber(blockNr BlockNumber) BlockNumberOrHash {
```

File: api/api_public_blockchain.go [hunk 6]
```diff
@@ -433,6 +461,26 @@ func FormatLogs(logs []vm.StructLog) []StructLogRes {
 	return formatted
 }
 
+// RPCOutputHeader converts the given header to the RPC output .
```
Comment (yoomee1313) [inline api/api_public_blockchain.go:464]:
```suggestion
// RPCOutputHeader converts the given header to the RPC output.
```

File: api/api_public_blockchain.go (hunk 2 above)
Comment (hyochan-brown) [reply api/api_public_blockchain.go:160]: I reflected it. Thank you always!

File: networks/rpc/types_test.go (hunk 4 above)
Comment (hyochan-brown) [reply networks/rpc/types_test.go]: I reflected it. 
Thank you yoomee1313!

File: api/api_public_blockchain.go (hunk 6 above)
Comment (hyochan-brown) [reply api/api_public_blockchain.go]: I reflected it. Thank you!

File: networks/rpc/types.go (hunk 5 above)
Comment (hyochan-brown) [reply networks/rpc/types.go]: There is no "New" in that method naming of ethereum.
But klaytn other code namings often have "New", so I added it for consistency. 
Thank you!

**Review by aidan-kwon — COMMENTED** [review]
@hyochan-brown Could you re-generate `console/jsre/deps/bindata.go` with the command in `console/jsre/deps/deps.go`? It should be updated when `console/jsre/deps/web3.js` is changed.

File: node/cn/api_backend.go [hunk 7]
```diff
@@ -96,6 +96,27 @@ func (b *CNAPIBackend) HeaderByNumber(ctx context.Context, blockNr rpc.BlockNumb
...
+		if header == nil {
+			return nil, fmt.Errorf("header for hash not found")
+		}
+		return header, nil
+	}
+	return nil, fmt.Errorf("invalid arguments; neither block nor hash specified")
+}
+
+func (b *CNAPIBackend) HeaderByHash(ctx context.Context, hash common.Hash) (*types.Header, error) {
+	return b.cn.blockchain.GetHeaderByHash(hash), nil
```
Comment (aidan-kwon) [inline node/cn/api_backend.go:117]: `GetHeaderByHash` could return `nil`. How about returning an error to the user instead of returning `nil, nil`?

File: node/cn/api_backend.go [hunk 8]
```diff
@@ -112,6 +133,31 @@ func (b *CNAPIBackend) BlockByNumber(ctx context.Context, blockNr rpc.BlockNumbe
...
+		return b.BlockByNumber(ctx, blockNr)
+	}
+	if hash, ok := blockNrOrHash.Hash(); ok {
+		block, err := b.BlockByHash(ctx, hash)
+		if err != nil {
+			return nil, err
+		}
+		if block == nil {
+			return nil, fmt.Errorf("header found, but block body is missing")
+		}
```
Comment (aidan-kwon) [inline node/cn/api_backend.go:155]: Unreachable case. `b.BlockByHash` doesn't return `(nil, nil)`.

File: console/web3ext/web3ext.go [hunk 9]
```diff
@@ -716,6 +716,17 @@ web3._extend({
...
+			name: 'getHeaderByNumber',
+			call: 'klay_getHeaderByNumber',
+			params: 1,
+			inputFormatter: [web3._extend.formatters.inputBlockNumberFormatter]
+		}),
+		new web3._extend.Method({
+			name: 'getHeaderByHash',
+			call: 'klay_getHeaderByHash',
+			params: 1
+		}),
```
Comment (aidan-kwon) [inline console/web3ext/web3ext.go:729]: `~NumberOrHash` API methods are implemented, but they are not introduced in web3ext. Why don't you add them?

**Review by aidan-kwon — COMMENTED** [review]
@hyochan-brown There are other methods requiring `rpc.BlockNumber` in `debug`, `governance`, or other namespaces. Could you apply `rpc.BlockNumberOrHash` to them later?

File: api/api_public_blockchain.go [hunk 10]
```diff
@@ -433,6 +453,26 @@ func FormatLogs(logs []vm.StructLog) []StructLogRes {
 	return formatted
 }
 
+// RPCOutputHeader converts the given header to the RPC output.
+func RPCOutputHeader(head *types.Header) (map[string]interface{}, error) {
```
Comment (aidan-kwon) [inline api/api_public_blockchain.go:457]: The output format is the same with json tags of `Header` struct. Why did you adds this converting function?

File: api/api_public_blockchain.go [hunk 11]
```diff
@@ -433,6 +453,26 @@ func FormatLogs(logs []vm.StructLog) []StructLogRes {
...
+	return map[string]interface{}{
+		"number":           (*hexutil.Big)(head.Number),
+		"parentHash":       head.ParentHash,
+		"logsBloom":        head.Bloom,
+		"stateRoot":        head.Root,
+		"reward":           head.Rewardbase,
+		"blockscore":       (*hexutil.Big)(head.BlockScore),
```
Comment (aidan-kwon) [inline api/api_public_blockchain.go:464]:
```suggestion
		"blockScore":       (*hexutil.Big)(head.BlockScore),
```

File: node/cn/api_backend.go (hunk 7 above)
Comment (hyochan-brown) [reply node/cn/api_backend.go:117]: I reflected it. Thank you!

File: node/cn/api_backend.go (hunk 8 above)
Comment (hyochan-brown) [reply node/cn/api_backend.go:155]: Oh,, I remove this code. Thank you!

File: api/api_public_blockchain.go (hunk 10 above)
Comment (hyochan-brown) [reply api/api_public_blockchain.go:457]: In some cases, there may be additional information that I want to include or hide, so I usually define the output model like rpcOutputBlock. But this case, it is unnecessary. I'll remove this code and use json tag of Header! 
Thank you aidan!

File: console/web3ext/web3ext.go (hunk 3 above)
Comment (hyochan-brown) [reply console/web3ext/web3ext.go:719]: I added getHeader!
Thank you!

File: api/api_public_blockchain.go (hunk 11 above)
Comment (hyochan-brown) [reply api/api_public_blockchain.go]: I modified it. Thank you.

**Comment by ehnuje** [comment]
@hyochan-brown Tests are failing. Could you please check?

File: console/jsre/deps/web3.js [hunk 12]
```diff
@@ -5340,15 +5340,19 @@ var Iban = require('../iban');
 var transfer = require('../transfer');
 
 var blockCall = function (args) {
-    return (utils.isString(args[0]) && args[0].indexOf('0x') === 0) ? "klay_getBlockByHash" : "klay_getBlockByNumber";
+    return (utils.isString(args[0]) && args[0].indexOf('0x') === 0 && args[0].length === 66) ? "klay_getBlockByHash" : "klay_getBlockByNumber";
```
Comment (jeongkyun-oh) [inline console/jsre/deps/web3.js:5343]: This is minor comment, but do you think thiat checking `args[0].length === 66` here is appropriate? For example, `klay_getBlock("0x0")` calls `klay_getBlockByNumber`. What do you think?

File: console/jsre/deps/web3.js [hunk 13]
```diff
@@ -5422,6 +5426,13 @@ var methods = function () {
         outputFormatter: formatters.outputBlockFormatter
     });
 
+    var getHeader = new Method({
+        name: 'getHeader',
+        call: headerCall,
+        params: 1,
+        inputFormatter: [formatters.inputBlockNumberFormatter]
```
Comment (jeongkyun-oh) [inline console/jsre/deps/web3.js:5433]: I have a question. Do you know what `function (val) { return !!val; }` is? `getBlock` has another parameter.

File: node/cn/api_backend.go [hunk 14]
```diff
@@ -91,11 +91,35 @@ func (b *CNAPIBackend) HeaderByNumber(ctx context.Context, blockNr rpc.BlockNumb
...
+func (b *CNAPIBackend) HeaderByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) (*types.Header, error) {
+	if blockNr, ok := blockNrOrHash.Number(); ok {
+		return b.HeaderByNumber(ctx, blockNr)
+	}
+	if hash, ok := blockNrOrHash.Hash(); ok {
+		header, err := b.HeaderByHash(ctx, hash)
+		if err != nil {
+			return nil, err
+		}
+		if header == nil {
```
Comment (jeongkyun-oh) [inline node/cn/api_backend.go:108]: Is it possible for the `headr` to be `nil`?

File: console/jsre/deps/web3.js (hunk 13 above)
Comment (hyochan-brown) [reply console/jsre/deps/web3.js:5433]: I was also curious when I wrote it. 
`getBlock` has one more parameter `fullTX` which type is boolean. 
I think `function (val) { return !!val; }` is ethereum style's boolean formatter...

File: networks/rpc/types.go [hunk 15]
```diff
@@ -170,3 +179,117 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
...
 }
+
+func (bn BlockNumber) Uint64() uint64 {
+	return (uint64)(bn)
+}
+
+type BlockNumberOrHash struct {
+	BlockNumber      *BlockNumber `json:"blockNumber,omitempty"`
+	BlockHash        *common.Hash `json:"blockHash,omitempty"`
+	RequireCanonical bool         `json:"requireCanonical,omitempty"`
```
[the remark this answers is not in the document]
Comment (hyochan-brown) [reply networks/rpc/types.go:190]: When we first discussed it, We have decided to have the same field to make compatibility of ethereum. @KimKyungup

File: console/jsre/deps/web3.js (hunk 12 above)
Comment (hyochan-brown) [reply console/jsre/deps/web3.js:5343]: Thanks for good review!
I have a question.
Could the hash value be `"0x0"` not `"0x000...000"`...?

File: console/jsre/deps/web3.js (hunk 13 above)
Comment (jeongkyun-oh) [reply console/jsre/deps/web3.js:5433]: Ah, thanks for explanation.

File: console/jsre/deps/web3.js (hunk 12 above)
Comment (jeongkyun-oh) [reply console/jsre/deps/web3.js:5343]: I meant wrong hash value. The correct value of hash is `0x000...000`, but if we put `0x0` to `getBlock` then it calls `getBlockByNumber`. What do you think about this?

File: api/api_public_blockchain.go [hunk 16]
```diff
@@ -442,7 +462,7 @@ func RpcOutputBlock(b *types.Block, td *big.Int, inclTx bool, fullTx bool) (map[
 		"logsBloom":        head.Bloom,
 		"stateRoot":        head.Root,
 		"reward":           head.Rewardbase,
-		"blockscore":       (*hexutil.Big)(head.BlockScore),
+		"blockScore":       (*hexutil.Big)(head.BlockScore),
```
Comment (aidan-kwon) [inline api/api_public_blockchain.go:465]: I agree that `blockScore` is the corrent name because `Header` struct has the same json tag. However, this change may affect users or BApps parsing with the `blockscore` keyword. Could you investigate APIs which use `blockscore` or `blockScore` currently?

File: node/cn/api_backend.go (hunk 14 above)
Comment (hyochan-brown) [reply node/cn/api_backend.go:108]: Oh.. header never be nil.
Thanks jk!

File: api/api_public_blockchain.go (hunk 16 above)
Comment (hyochan-brown) [reply api/api_public_blockchain.go]: I'll restore it and discuss it at the weekly meeting.
Thanks Aidan!

File: console/jsre/deps/web3.js (hunk 12 above)
Comment (hyochan-brown) [reply console/jsre/deps/web3.js:5343]: Not only blockCall, BlockNumberOrHash is also divided based on whether it's length is 66. (Ethereum same)
So I think .. even if the user intended blockhash, `0x0` should be considered a blocknumber in code.

File: api/api_public_debug.go [hunk 17]
```diff
@@ -56,19 +57,24 @@ func (api *PublicDebugAPI) GetBlockRlp(ctx context.Context, number uint64) (stri
...
+		blockNumberOrHashString, _ := blockNrOrHash.NumberOrHashString()
+		return "", fmt.Errorf("block %v not found", blockNumberOrHashString)
 	}
 	return spew.Sdump(block), nil
 }
 
 // SeedHash retrieves the seed hash of a block.
-func (api *PublicDebugAPI) SeedHash(ctx context.Context, number uint64) (string, error) {
-	block, _ := api.b.BlockByNumber(ctx, rpc.BlockNumber(number))
+func (api *PublicDebugAPI) SeedHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) (string, error) {
```
Comment (aidan-kwon) [inline api/api_public_debug.go:70]: 1. The input type is changed to support blockHash but the API doesn't support blockHash.
2. This API is for gxhash(PoW) consensus algorithm which is not maintained. I think we better deprecate this API.

File: networks/rpc/types.go [hunk 18]
```diff
@@ -138,6 +140,13 @@ const (
 // - an invalid block number error when the given argument isn't a known strings
 // - an out of range error when the given block number is either too little or too large
 func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
+	var decimalInput uint64
+	err := json.Unmarshal(data, &decimalInput)
+	if err == nil {
+		*bn = BlockNumber(decimalInput)
+		return nil
+	}
```
Comment (aidan-kwon) [inline networks/rpc/types.go:148]: Could you explain why this logic is added though a similar logic exist in the below?

File: networks/rpc/types.go [hunk 19]
```diff
@@ -170,3 +179,117 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
...
+
+	var decimalInput uint64
+	err = json.Unmarshal(data, &decimalInput)
+	if err == nil {
+		bn := BlockNumber(decimalInput)
+		bnh.BlockNumber = &bn
+		return nil
+	}
```
Comment (aidan-kwon) [inline networks/rpc/types.go:213]: Could you explain why this logic is added though a similar logic exist in the below?

File: networks/rpc/types.go [hunk 20]
```diff
@@ -170,3 +179,117 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
...
+}
+
+func (bnh *BlockNumberOrHash) Hash() (common.Hash, bool) {
+	if bnh.BlockHash != nil {
+		return *bnh.BlockHash, true
+	}
+	return common.Hash{}, false
+}
+
+func (bnh *BlockNumberOrHash) NumberOrHashString() (string, bool) {
```
Comment (aidan-kwon) [inline networks/rpc/types.go:271]: The name sounds number has priority when both number and hash values are assigned. However, hash has more priority than number in the following logic. Did you intended it?

File: networks/rpc/types_test.go [hunk 21]
```diff
@@ -42,7 +43,7 @@ func TestBlockNumberJSONUnmarshal(t *testing.T) {
 		6:  {`"0x12"`, false, BlockNumber(18)},
 		7:  {`"0x7fffffffffffffff"`, false, BlockNumber(math.MaxInt64)},
 		8:  {`"0x8000000000000000"`, true, BlockNumber(0)},
-		9:  {"0", true, BlockNumber(0)},
+		9:  {"0x0", true, BlockNumber(0)},
```
Comment (aidan-kwon) [inline networks/rpc/types_test.go:46]: Why did you change this value?

File: api/api_public_debug.go (hunk 17 above)
Comment (hyochan-brown) [reply api/api_public_debug.go:70]: 1. Oh.. I misunderstood this API as only for blockNumber. It's possible to support blockHash.
```
func (api *PublicDebugAPI) SeedHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) (string, error) {	
	block, err := api.b.BlockByNumberOrHash(ctx, blockNr)	
	if err != nil {		
	   blockNumberOrHashString, _ := blockNrOrHash.NumberOrHashString()
	   return "", fmt.Errorf("block %v not found", blockNumberOrHashString)
	}	
	return fmt.Sprintf("0x%x", gxhash.SeedHash(uint64(block.header.number))), nil}
}
```
2. However, as you said, it seems unnecessary API, so I will deprecate this API.

Thanks! Aidan.

File: networks/rpc/types.go [hunk 22]
```diff
@@ -170,3 +179,117 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
...
+	if bnh.BlockHash != nil {
+		return bnh.BlockHash.String(), true
+	}
+	if bnh.BlockNumber != nil {
+		return fmt.Sprintf("#%v", bnh.BlockNumber.Int64()), true
+	}
+	return "", false
+}
+
+func NewBlockNumberOrHashWithNumber(blockNr BlockNumber) BlockNumberOrHash {
```
Comment (aidan-kwon) [inline networks/rpc/types.go:281]: How about using the same naming rule with Ethereums?

File: networks/rpc/types.go [hunk 23]
```diff
@@ -170,3 +179,117 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
...
+	return BlockNumberOrHash{
+		BlockNumber:      &blockNr,
+		BlockHash:        nil,
+		RequireCanonical: false,
+	}
+}
+
+func NewBlockNumberOrHashWithHash(hash common.Hash, canonical bool) BlockNumberOrHash {
```
Comment (aidan-kwon) [inline networks/rpc/types.go:289]: How about using the same naming rule with Ethereums?

File: node/cn/api.go [hunk 24]
```diff
@@ -215,18 +215,20 @@ func NewPublicDebugAPI(cn *CN) *PublicDebugAPI {
...
+		block = api.cn.APIBackend.CurrentBlock()
 	} else {
-		block = api.cn.blockchain.GetBlockByNumber(uint64(blockNr))
+		block, err = api.cn.APIBackend.BlockByNumberOrHash(ctx, blockNrOrHash)
 	}
-	if block == nil {
-		return state.Dump{}, fmt.Errorf("block #%d not found", blockNr)
+	if err != nil {
+		blockNrOrHashString, _ := blockNrOrHash.NumberOrHashString()
+		return state.Dump{}, fmt.Errorf("block %v not found", blockNrOrHashString)
```
Comment (aidan-kwon) [inline node/cn/api.go:231]: Why did you handle error out of `else` branch?

File: networks/rpc/types.go (hunk 22 above)
Comment (hyochan-brown) [reply networks/rpc/types.go:281]: There is no "New" in that method naming of ethereum.
But klaytn other code namings often have "New", so I added it for consistency...
I don't know which is more important, consistency or compatibility.
Frankly, I don't know the team's convention yet. :cry: :cry:

File: networks/rpc/types.go (hunk 18 above)
Comment (hyochan-brown) [reply networks/rpc/types.go:148]: BlockNumber(orHash) supports only hex, decimal is not supported.
So if API's input is decimal, param type cannot be BlockNumber. 
I think it was inconsistent. 
This logic is to support decimal(uint64) input in BlockNumber(OrHash).

File: networks/rpc/types.go (hunk 20 above)
Comment (hyochan-brown) [reply networks/rpc/types.go:271]: Both number and hash cannot be assigned. I thought the order would not matter.
But I'll change the order to match the naming.
Thanks!

File: node/cn/api.go (hunk 24 above)
Comment (hyochan-brown) [reply node/cn/api.go]: Oh,, It's my mistake. I moved error check logic in else branch 
Thanks Aidan!

File: networks/rpc/types_test.go (hunk 21 above)
Comment (hyochan-brown) [reply networks/rpc/types_test.go]: `9:  {"0", true, BlockNumber(0)},`
This code is fail test when a decimal input comes.
I've changed it because I think it's unnecessary.

Now thought, this can  be as a successful unit test.
I added some unit tests including this test .
Thanks!

File: api/api_public_blockchain.go [hunk 25]
```diff
@@ -150,6 +150,26 @@ func (s *PublicBlockChainAPI) GetAccount(ctx context.Context, address common.Add
...
 }
 
+// GetHeaderByNumber returns the requested canonical block header.
+func (s *PublicBlockChainAPI) GetHeaderByNumber(ctx context.Context, number rpc.BlockNumber) (*types.Header, error) {
+	header, err := s.b.HeaderByNumber(ctx, number)
+	if err != nil {
+		return nil, err
+	}
+	return header, nil
+
```
Comment (yoomee1313) [inline api/api_public_blockchain.go:160]: @hyochan-brown I think it can be simplified like this. What do you think? Also, there seems to be more code similar to this.
```suggestion
	return s.b.HeaderByNumber(ctx, number)
```

File: networks/rpc/types.go (hunk 22 above)
Comment (yoomee1313) [reply networks/rpc/types.go:281]: I think it was my suggestion.. sorry for the confusion! If ethereum compatibility is more important, it would be better to go back.

File: networks/rpc/types_test.go [hunk 26]
```diff
@@ -42,14 +43,19 @@ func TestBlockNumberJSONUnmarshal(t *testing.T) {
 		6:  {`"0x12"`, false, BlockNumber(18)},
 		7:  {`"0x7fffffffffffffff"`, false, BlockNumber(math.MaxInt64)},
 		8:  {`"0x8000000000000000"`, true, BlockNumber(0)},
-		9:  {"0", true, BlockNumber(0)},
+		9:  {`"0"`, true, BlockNumber(0)},
```
Comment (yoomee1313) [inline networks/rpc/types_test.go:46]: Can I ask why this is fixed?

Comment (hyochan-brown) [reply networks/rpc/types_test.go:46]: original - `9:  {"0", true, BlockNumber(0)},`
second field is `mustFail`.
This test code is decimal input fail test.
After BlockNumber cover decimal, 
this test cannot be passed (because not fail)
So I change second field `true` to `false`.
And it's moved to no.17 for test consistency. (1~16 : String hex test / 17~21 : decimal test)
now - `17:  {"0", false, BlockNumber(0)},`

`9:  {`"0"`, true, BlockNumber(0)},`
this is new test for check string decimal fail.

File: api/api_public_blockchain.go (hunk 25 above)
Comment (hyochan-brown) [reply api/api_public_blockchain.go]: Thank you for the good review.
I reflected it!

File: networks/rpc/types_test.go (hunk 26 above)
Comment (yoomee1313) [reply networks/rpc/types_test.go:46]: Thanks for the kind explanation. It's really helpful for understanding rest of the test code :)

File: api/api_public_blockchain.go [hunk 27]
```diff
@@ -150,6 +150,18 @@ func (s *PublicBlockChainAPI) GetAccount(ctx context.Context, address common.Add
...
+// GetHeaderByNumber returns the requested canonical block header.
+func (s *PublicBlockChainAPI) GetHeaderByNumber(ctx context.Context, number rpc.BlockNumber) (*types.Header, error) {
+	return s.b.HeaderByNumber(ctx, number)
+
+}
+
+// GetHeaderByHash returns the requested header by hash.
+func (s *PublicBlockChainAPI) GetHeaderByHash(ctx context.Context, hash common.Hash) (*types.Header, error) {
+	return s.b.HeaderByHash(ctx, hash)
+
```
Comment (yoomee1313) [inline api/api_public_blockchain.go:162]: Please remove a spacebar :)

File: api/api_public_blockchain.go [hunk 28]
```diff
@@ -150,6 +150,18 @@ func (s *PublicBlockChainAPI) GetAccount(ctx context.Context, address common.Add
 	return serAcc, state.Error()
 }
 
+// GetHeaderByNumber returns the requested canonical block header.
+func (s *PublicBlockChainAPI) GetHeaderByNumber(ctx context.Context, number rpc.BlockNumber) (*types.Header, error) {
+	return s.b.HeaderByNumber(ctx, number)
+
```
[the remark this answers is not in the document]
Comment (hyochan-brown) [reply api/api_public_blockchain.go]: Oh.. I remove this.
Thanks yoomee1313!

File: networks/rpc/types.go (hunk 18 above)
Comment (aidan-kwon) [reply networks/rpc/types.go:148]: Current
`1` --marshal & BlockNumber.UnmarshlJSON--> fail
`0x1` --marshal & BlockNumber.UnmarshlJSON--> fail
`"1"` --marshal & BlockNumber.UnmarshlJSON-->  fail 
`"0x1"` --marshal & BlockNumber.UnmarshlJSON-->  success 
`"latest"` --marshal & BlockNumber.UnmarshlJSON-->  success 

Your suggestion 
`1` --marshal & BlockNumber.UnmarshlJSON-->  success
`0x1` --marshal & BlockNumber.UnmarshlJSON-->  success
`"1"` --marshal & BlockNumber.UnmarshlJSON-->  fail 
`"0x1"` --marshal & BlockNumber.UnmarshlJSON-->  success 
`"latest"` --marshal & BlockNumber.UnmarshlJSON-->  success 

@hyochan-brown Is it the same with your intention? I think it is little bit confusing.

File: networks/rpc/types.go (hunk 23 above)
Comment (aidan-kwon) [reply networks/rpc/types.go:289]: @hyochan-brown Could you answer me?

Comment (aidan-kwon) [reply networks/rpc/types.go:289]: I saw the question above. Ok. Let it go.

File: networks/rpc/types.go (hunk 18 above)
Comment (aidan-kwon) [reply networks/rpc/types.go:148]: If you intended it, `TestBlockNumberJSONUnmarshal` need to be changed to test `number` type input. Currently, only `string` type inputs have beeing tested.

File: node/cn/api.go [hunk 29]
```diff
@@ -215,18 +215,20 @@ func NewPublicDebugAPI(cn *CN) *PublicDebugAPI {
...
 	} else {
-		block = api.cn.blockchain.GetBlockByNumber(uint64(blockNr))
-	}
-	if block == nil {
-		return state.Dump{}, fmt.Errorf("block #%d not found", blockNr)
+		block, err = api.cn.APIBackend.BlockByNumberOrHash(ctx, blockNrOrHash)
+		if err != nil {
+			blockNrOrHashString, _ := blockNrOrHash.NumberOrHashString()
+			return state.Dump{}, fmt.Errorf("block %v not found", blockNrOrHashString)
+		}
```
Comment (aidan-kwon) [inline node/cn/api.go:231]: `nil` can be assigned by `api.cn.APIBackend.CurrentBlock()`. Could you let `block == nil` checking be stayed?

File: networks/rpc/types.go (hunk 18 above)
Comment (hyochan-brown) [reply networks/rpc/types.go:148]: Yes. I intended it. 
Because current BlockNumber cannot unmarshal decimal input, so some API's block number type is integer, not BlockNumber.
I wanted to add consistency in API spec.
I added some test in `TestBlockNumberJSONUnmarshal`.(17-21)
`"n"` is n (number), `'"n"'` is "n" (string) in this test.

Comment (aidan-kwon) [reply networks/rpc/types.go:148]: @hyochan-brown `"n"` is not a number type. It is a string type having a number value. 
```
json.Marshal(1) == []byte{49}
json.Marsahl("1") == []byte{34, 49, 34}
```
Since `TestBlockNumberJSONUnmarshal` doesn't  marshal itsel and use string input, it cannot test number types such int.

Comment (aidan-kwon) [reply networks/rpc/types.go:148]: @hyochan-brown My mistake. You're right. `json.Marshal(1)  == []byte("1")`

**Review by aidan-kwon — APPROVED** [review]
LGTM

**Review by yoomee1313 — APPROVED** [review]

**Review by jeongkyun-oh — APPROVED** [review]
